src/lib/ledger_catchup/normal_catchup.ml

```diff
@@ -173,24 +173,82 @@ let rec fold_until ~(init : 'accum)
       | Continue_or_stop.Continue init ->
           fold_until ~init ~f ~finish xs )
 
+let find_map_ok l ~f =
+  Deferred.repeat_until_finished (l, []) (fun (l, errors) ->
+      match l with
+      | [] ->
+          let errors = List.rev errors in
+          Deferred.return @@ `Finished (Error errors)
+      | hd :: tl ->
+          Deferred.map (f hd) ~f:(function
+            | Error current_error ->
+                `Repeat (tl, current_error :: errors)
+            | Ok result ->
+                `Finished (Ok result) ) )
+
+type download_state_hashes_error =
+  [ `Peer_moves_too_fast of Error.t
+  | `No_common_ancestor of Error.t
+  | `Failed_to_download_merkle_path of Error.t
```

Comment:
change `Failed_to_download_merkle_path`  to `Failed_to_download_transition_chain_proof` and `Invalid_merkle_path` to `Invalid_transition_chain_proof`?